video/out/opengl/hwdec_surfacetexture.c

```diff
@@ -0,0 +1,432 @@
+    gl->TexParameteri(GL_TEXTURE_EXTERNAL_OES, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+    gl->TexParameteri(GL_TEXTURE_EXTERNAL_OES, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+    gl->BindTexture(GL_TEXTURE_EXTERNAL_OES, 0);
+
+    MP_JNI_CALL_VOID(o->texture, SurfaceTexture.attachToGLContext, p->gl_texture);
+    if (MP_JNI_EXCEPTION_LOG(mapper) < 0) return -1;
+
+    return 0;
+}
+
+static void mapper_uninit(struct ra_hwdec_mapper *mapper)
+{
+    struct priv *p = mapper->priv;
+    struct priv_owner *o = mapper->owner->priv;
+    GL *gl = ra_gl_get(mapper->ra);
+    JNIEnv *env = MP_JNI_GET_ENV(mapper);
+
+    if (p->gl_texture) {
+        MP_JNI_CALL_VOID(o->texture, SurfaceTexture.detachFromGLContext);
+        MP_JNI_EXCEPTION_LOG(mapper);
+        gl->DeleteTextures(1, &p->gl_texture);
+    }
+
+    if (p->listener) {
+        jmethodID set_native_ptr_id = (*env)->GetMethodID(env, p->listener_class, "setNativePtr", "(J)V");
+        MP_JNI_EXCEPTION_LOG(mapper);
+        (*env)->CallVoidMethod(env, p->listener, set_native_ptr_id, (jlong)0);
+        MP_JNI_EXCEPTION_LOG(mapper);
+
+        MP_JNI_CALL_VOID(o->texture, SurfaceTexture.setOnFrameAvailableListener, NULL);
+        MP_JNI_EXCEPTION_LOG(mapper);
+
+        (*env)->DeleteGlobalRef(env, p->listener);
+        p->listener = NULL;
+    }
+
+    (*env)->DeleteGlobalRef(env, p->matrix);
+
+    pthread_mutex_destroy(&p->lock);
+    pthread_cond_destroy(&p->cond);
+}
+
+static void mapper_unmap(struct ra_hwdec_mapper *mapper)
+{
+    ra_tex_free(mapper->ra, &mapper->tex[0]);
+}
+
+static int mapper_map(struct ra_hwdec_mapper *mapper)
+{
+    struct priv *p = mapper->priv;
+    struct priv_owner *o = mapper->owner->priv;
+    JNIEnv *env = MP_JNI_GET_ENV(mapper);
+
+    if (!p->gl_texture)
+        return -1;
+
+    AVMediaCodecBuffer *buffer = (AVMediaCodecBuffer *)mapper->src->planes[3];
+    int frame_available = 0;
+
+    pthread_mutex_lock(&p->lock);
+    p->on_frame_available = 0;
+    av_mediacodec_release_buffer(buffer, 1);
+    if (p->listener) {
+        struct timespec ts = mp_rel_time_to_timespec(0.030);
+        int ret = pthread_cond_timedwait(&p->cond, &p->lock, &ts);
+        if (ret < 0)
+            MP_WARN(mapper, "Failed to wait on signal: %s\n", mp_strerror(ret));
+        frame_available = p->on_frame_available;
+    } else {
+        frame_available = 1;
+    }
+    pthread_mutex_unlock(&p->lock);
+
+    if (!frame_available)
+        MP_WARN(mapper, "no frame!\n");
+
+    MP_JNI_CALL_VOID(o->texture, SurfaceTexture.updateTexImage);
+    if (MP_JNI_EXCEPTION_LOG(mapper->owner)) return -1;
+
+    MP_JNI_CALL_VOID(o->texture, SurfaceTexture.getTransformMatrix, p->matrix);
+    if (MP_JNI_EXCEPTION_LOG(mapper->owner)) return -1;
+
+    (*env)->GetFloatArrayRegion(env, p->matrix, 0, 16, p->transform);
+    if (MP_JNI_EXCEPTION_LOG(mapper->owner)) return -1;
+
+    mapper->transform = (struct gl_transform){
```

Comment:
I doubt the transform is wrong. let M be the matrix from getTransformMatrix, F be the flip matrix, M maps sampler point (x,y,..) to texture point F*(x,y, ..), so the result matrix should be M*F^-1=M*F. But i don't know how mpv uses tranform in shader.

Do you have an prebuilt apk for testing?

Reply:
mediacodec can not be enabled in this apk